### PR TITLE
RF `publish` in many ways

### PR DESCRIPTION
**Pull request by mih**
- [x] `publish` need to become a new-style command using path annotation
- [x] #1495 --since might be not in effect at all
- [x] #1480 bug: publish should publish (annex copy) data first before pushing
- [x] #1426 bug: doesn't publish
- [x] remove horrible `Dataset.get_containing_subdataset()`

### CODE REVIEW

**Comment by mih** [comment]
Nice segfault in tests, awesome!

File: datalad/distribution/tests/test_publish.py [hunk 1]
```diff
@@ -268,24 +293,22 @@ def test_publish_with_data(origin, src_path, dst_path, sub1_pub, sub2_pub, dst_c
     source.repo.add_remote("target", dst_path)
 
     # now, set up targets for the submodules:
-    sub1_target = GitRepo(sub1_pub, create=True)
+    # the need to be annexes, because we want to be able to copy data to them
+    # further down
+    sub1_target = AnnexRepo(sub1_pub, create=True)
```
Comment (yarikoptic) [inline datalad/distribution/tests/test_publish.py]: I think that was on purpose -- is should be ok to have superdataset to be a pure git, but subdatasets - annex. and publish the whole collection

Comment (mih) [inline datalad/distribution/tests/test_publish.py]: True, but this is about the subdataset.

**Comment by mih** [comment]
The current state of this PR completes the conversion to new-style commands in that all old helpers could be removed. The few remaining commands are even older and never used the old-new helpers, or it does not make sense for them to use such code at all. Here is the list:

- `add-archive-content`
- `add-sibling`
- `aggregate-metadata`
- `crawl-init`
- `crawl`
- `create-sibling-github`
- `create-test-dataset`
- `download-url`
- `export`
- `ls`
- `move`
- `rewrite-urls`
- `search`
- `sshrun`
- `test`

**Comment by yarikoptic** [comment]
example fails now @mih

**Comment by yarikoptic** [comment]
so wrapt is hunting us down .... note that on travis it installs it from pypi the most recent 1.10.10 whenever in debian we have only 1.9.0 ... might be worth exploring the downgrade for a quick and dirty workaround...?

**Comment by mih** [comment]
And it starts to fail from Debian 8 (with wrapt 1.8, works) to Debian 9
(with wrapt 1.9, broken).

**Comment by yarikoptic** [comment]
probably picked up somewhere else among failures but here is what I get when testing within docker using python-dbg (chasing that gc issue... heh):
```
======================================================================
ERROR: datalad.distribution.tests.test_publish.test_invalid_call
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/tmp/datalad/datalad/tests/utils.py", line 762, in newfunc
    t(*(arg + (uri,)), **kw)
  File "/tmp/datalad/datalad/tests/utils.py", line 592, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/tmp/datalad/datalad/distribution/tests/test_publish.py", line 47, in test_invalid_call
    ds.uninstall('subm 1', check=False)
  File "/usr/local/lib/python2.7/dist-packages/wrapt/wrappers.py", line 561, in __call__
    args, kwargs)
  File "/tmp/datalad/datalad/distribution/dataset.py", line 403, in apply_func
    return f(**kwargs)
  File "/usr/local/lib/python2.7/dist-packages/wrapt/wrappers.py", line 522, in __call__
    args, kwargs)
  File "/tmp/datalad/datalad/interface/utils.py", line 511, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/wrapt/wrappers.py", line 522, in __call__
    args, kwargs)
  File "/tmp/datalad/datalad/interface/utils.py", line 499, in return_func
    results = list(results)
  File "/tmp/datalad/datalad/interface/utils.py", line 490, in generator_func
    msg="Command did not complete successfully")
IncompleteResultsError: Command did not complete successfully [{'status': 'error', 'orig_request': 'subm 1', 'registered_subds': True, 'raw_input': True, 'parentds': '/tmp/datalad_temp_testrepo_vY4iwT', 'action': 'uninstall', 'path': '/tmp/datalad_temp_testrepo_vY4iwT/subm 1', 'message': 'refusing to uninstall current or parent directory', 'type': 'dataset', 'refds': '/tmp/datalad_temp_testrepo_vY4iwT'}]

```

**Comment by mih** [comment]
re uninstall error: a number of these were fixed in #1575

**Comment by mih** [comment]
Published a rebase on current master. Now with fewer commits. Let's see how it performs. If it fails again, I will factor out a few more.

**Comment by mih** [comment]
FTR: This failure is not specific to the test where the segfault usually occurs. When that one is disabled the next one fails: https://travis-ci.org/datalad/datalad/jobs/244968746

**Comment by mih** [comment]
FTR: Moving the publish test to an ealier point in the test run does NOT change the failure pattern. That makes a general leakage as the cause more unlikely, and something publish-related more likely.

**Comment by mih** [comment]
Rebased on dev-master.

**Pull request merged**